### PR TITLE
NKAI: composite pathfinder actions and fix for guarded bordergate

### PR DESCRIPTION
**Pull request by nullkiller**
AI endlessly try to bypass back-guarded questguard bordergate
For instance M|M|G|H or M|Q|H. Guard covers border gate / questguard and battle replaces QuestAction thus making quest invisible to AI.

### CODE REVIEW

File: AI/Nullkiller/Pathfinding/AINodeStorage.h [hunk 1]
```diff
@@ -10,8 +10,8 @@
 
 #pragma once
 
-#define NKAI_PATHFINDER_TRACE_LEVEL 0
-#define NKAI_TRACE_LEVEL 0
+#define NKAI_PATHFINDER_TRACE_LEVEL 1
+#define NKAI_TRACE_LEVEL 2
```
Comment (IvanSavenko) [inline AI/Nullkiller/Pathfinding/AINodeStorage.h:14]: Intended?

Comment (nullkiller) [reply AI/Nullkiller/Pathfinding/AINodeStorage.h:14]: No, will rollback

File: AI/Nullkiller/Pathfinding/Actions/SpecialAction.cpp [hunk 2]
```diff
@@ -27,4 +27,64 @@ void SpecialAction::execute(const CGHeroInstance * hero) const
...
+	}
+}
+
+std::string CompositeAction::toString() const
+{
+	std::string result = "";
+
+	for(auto part : parts)
+	{
+		result += part->toString();
```
Comment (Nordsoft91) [inline AI/Nullkiller/Pathfinding/Actions/SpecialAction.cpp:79]: Maybe add separator?

File: AI/Nullkiller/Pathfinding/Actions/SpecialAction.cpp [hunk 3]
```diff
@@ -27,4 +27,64 @@ void SpecialAction::execute(const CGHeroInstance * hero) const
...
+	}
+
+	return result;
+}
+
+const CGObjectInstance * CompositeAction::targetObject() const
+{
+	return parts.front()->targetObject();
```
Comment (Nordsoft91) [inline AI/Nullkiller/Pathfinding/Actions/SpecialAction.cpp:87]: Let's add some check for !parts.empty(), at least assert

**Pull request merged**